nerd/data/dataset.go

```diff
@@ -0,0 +1,187 @@
+package data
+
+import (
+	"bufio"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	//newline is the separator for Keys in the metadata file.
+	newline  = '\n'
+	newlineS = "\n"
+
+	//MetadataObjectKey is the key of the S3 object that holds the metadata of a dataset.
+	MetadataObjectKey = "metadata"
+)
+
+//Key is the identifier of a chunk of data.
+type Key [sha256.Size]byte
+
+//ToString returns the string representation of a key.
+func (k Key) ToString() string {
+	return fmt.Sprintf("%x", k)
+}
+
+//ZeroKey is an empty key.
+var ZeroKey = Key{}
+
+//KeyReadWriter is an interface that wraps the read and write methods for Keys.
+//
+//Read always returns a Key unless no more keys are available. In this case Read returns ZeroKey and an io.EOF error.
+//Write writes a Key and returns an error if something goes wrong.
+type KeyReadWriter interface {
+	ReadKey() (Key, error)
+	WriteKey(Key) error
+}
+
+//Metadata describes a dataset. It contains a header with different properties of the dataset and a
+//KeyReadWriter which is used to keep track of the list of Keys (chunks) of the dataset.
+type Metadata struct {
+	Header *MetadataHeader
+	KeyReadWriter
+}
+
+//MetadataHeader is the header of the dataset metadata.
+type MetadataHeader struct {
+	Created time.Time `json:"created_at"`
```

Comment:
these times are probably set by local machine's clock? UTC? its probably ok, we just need to be careful that they are not used for any logic (clock skews and what not)